rllib/utils/exploration/parameter_noise.py

```diff
@@ -0,0 +1,243 @@
+                randomly (see [1]).
+            framework (Optional[str]): One of None, "tf", "torch".
+        """
+        assert framework is not None
+        super().__init__(action_space, framework=framework, **kwargs)
+
+        self.stddev = get_variable(
+            initial_stddev, framework=self.framework, tf_name="stddev")
+        # Our noise to be added to the weights. Each item in `self.noise`
+        # corresponds to one Model variable and holding the Gaussian noise to
+        # be added to that variable (weight).
+        self.noise = None
+        # The weight variables of the Model where noise should be applied to.
+        # This excludes any variable, whose name contains "LayerNorm" (those
+        # are BatchNormalization layers, which should not be perturbed).
+        self.model_variables = None
+
+        self.sub_exploration = None
+        # For discrete action spaces, use an underlying EpsilonGreedy with a
+        # special schedule.
+        if isinstance(self.action_space, Discrete):
+            self.sub_exploration = EpsilonGreedy(
+                self.action_space,
+                epsilon_schedule={
+                    "type": "PiecewiseSchedule",
+                    "endpoints": [(0, 1.0), (random_timesteps, 0.01)],
+                    "outside_value": 0.01
+                })
+        # TODO(sven): Implement for any action space.
+        else:
+            raise NotImplementedError
+
+    @override(Exploration)
+    def forward(self,
+                model,
+                obs_batch,
+                *,
+                state_batches=None,
+                seq_lens=None,
+                prev_action_batch=None,
+                prev_reward_batch=None,
+                explore=True,
+                is_training=True):
+        # Initialize all noise_vars.
+        if self.noise is None:
+            self.model_variables = [
+                v for v in model.variables() if "LayerNorm" not in v.name
+            ]
+            self.noise = []
+            for var in self.model_variables:
+                self.noise.append(
+                    get_variable(
+                        np.zeros(var.shape, dtype=np.float32),
+                        framework=self.framework,
+                        tf_name=var.name.split(":")[0] + "_noisy"))
+
+        input_dict = {SampleBatch.CUR_OBS: obs_batch}
+        if self.framework == "tf" and not tf.executing_eagerly():
+            return self._tf_forward_op(explore, model, input_dict,
+                                       state_batches, seq_lens)
+        else:
+            # Exploration: Normal behavior.
+            if explore:
+                return model(input_dict, state_batches, seq_lens)
+
+            # No exploration: Revert noise, forward-pass, then re-apply noise.
+            else:
+                self._remove_noise(model)
+                out = model(input_dict, state_batches, seq_lens)
+                self._apply_noise(model)
```

Comment:
Is the only reason to override forward to wrap it in remove_noise / apply_noise?

If so, I think we can add less intrusive callbacks like `before_forward_pass` `after_forward_pass`.

Reply:
(overall, I would prefer a small set of APIs that compose together rather than a monolithic override (unless we had no choice of course)).

Reply:
(overall, I would prefer a smaller set of APIs that compose together rather than a monolithic override (unless we had no choice of course)).

Reply:
Yeah, maybe that's a less intrusive approach. I'll check whether this is possible. For param-noise, definitely, and it won't be much work to change this as it's only used in one Exploration class so far and the policy can just do again what's currently in `(base)Exploration.forward()`.